src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,105 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double GROW_INDEX = 1.5;
+    private T[] elementData;
+    private int size;
+
+    public ArrayList() {
+        this.elementData = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == elementData.length) {
+            grow();
+        }
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index: " + index);
+        }
+        if (size == elementData.length) {
+            grow();
+        }
+        System.arraycopy(elementData, index, elementData, index + 1, size - index);
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        int newSize = size + list.size();
+        if (newSize > elementData.length) {
+            grow();
+        }
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= 0 && index < size) {
+            return elementData[index];
+        }
+        throw new ArrayListIndexOutOfBoundsException("Index: " + index + ", is invalid ");
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index >= 0 && index < size) {
+            elementData[index] = value;
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("Index: " + index + ", is invalid ");
+        }
```

Comment:
avoid using if-else construction. 
```suggestion
        checkIndex(index); 
        elementData[index] = value;
```
And create private method:
```
private void checkIndex(int index) {
        if (index < 0 || index >= size) {
           throw new ArrayListIndexOutOfBoundsException("Index: " + index + ", is invalid ");
        }
    }
```